Sources/AsyncAlgorithms/Merge2StateMachine.swift

```diff
@@ -0,0 +1,162 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+struct Merge2StateMachine<Base1: AsyncSequence, Base2: AsyncSequence>: Sendable where Base1.AsyncIterator: Sendable, Base2.AsyncIterator: Sendable, Base1.Element: Sendable, Base2.Element: Sendable {
+  typealias Element1 = Base1.Element
+  typealias Element2 = Base2.Element
+
+  let iter1TerminatesOnNil : Bool
+  let iter2terminatesOnNil : Bool
+
+  enum Partial: @unchecked Sendable {
+    case first(Result<Element1?, Error>, Base1.AsyncIterator)
+    case second(Result<Element2?, Error>, Base2.AsyncIterator)
+  }
+
+  enum Either {
+    case first(Base1.Element)
+    case second(Base2.Element)
+  }
+
+  var state: (PartialIteration<Base1.AsyncIterator, Partial>, PartialIteration<Base2.AsyncIterator, Partial>)
+
+  init(_ iterator1: Base1.AsyncIterator, terminatesOnNil iter1TerminatesOnNil: Bool = false, _ iterator2: Base2.AsyncIterator, terminatesOnNil iter2terminatesOnNil: Bool = false) {
+    self.iter1TerminatesOnNil = iter1TerminatesOnNil
+    self.iter2terminatesOnNil = iter2terminatesOnNil
+    state = (.idle(iterator1), .idle(iterator2))
+  }
+
+  mutating func apply(_ task1: Task<Merge2StateMachine<Base1, Base2>.Partial, Never>?, _ task2: Task<Merge2StateMachine<Base1, Base2>.Partial, Never>?) async rethrows -> Either? {
+    switch await Task.select([task1, task2].compactMap ({ $0 })).value {
```

Comment:
do you think there is a way we can approach this without `Task.select`? That is the other perf bottle neck (which could be either solved by not needing it, or solved by improving it at a language level)

Reply:
Same here, not my code just copied over.